crates/ursa-gateway/src/indexer/mod.rs

```diff
@@ -1,52 +1,182 @@
 pub mod model;
 
-use crate::indexer::model::IndexerResponse;
-use anyhow::{anyhow, bail, Error};
-use libp2p::multiaddr::Protocol;
 use std::net::SocketAddrV4;
-use tracing::warn;
 
-// Chooses a provider and returns its addresses.
-pub fn choose_provider(indexer_response: IndexerResponse) -> Result<Vec<SocketAddrV4>, Error> {
-    // We expect that MultihashResults will have at most 1 element.
+use anyhow::{bail, Context, Result};
+use hyper::{body, client::HttpConnector, Body, Request, Uri};
+use hyper_tls::HttpsConnector;
+use jsonrpc_v2::{Id, V2};
+use libp2p::multiaddr::Protocol;
+use serde::Deserialize;
+use serde_json::{from_slice, json};
+use tracing::{debug, error};
+
+use crate::indexer::model::IndexerResponse;
+
+type Client = hyper::client::Client<HttpsConnector<HttpConnector>, Body>;
+
+pub struct Indexer {
+    cid_url: String,
+    client: Client,
+}
+
+impl Indexer {
+    pub fn new(cid_url: String, client: Client) -> Self {
+        Self { cid_url, client }
+    }
+
+    pub async fn query(&self, cid: &str) -> Result<Vec<SocketAddrV4>> {
+        let endpoint = format!("{}/{cid}", self.cid_url);
+        let uri = match endpoint.parse::<Uri>() {
+            Ok(uri) => uri,
+            Err(e) => {
+                error!("Error parsed uri: {endpoint} {e}.");
+                bail!("Error parsed uri: {endpoint}.")
+            }
+        };
+
+        let resp = match self.client.get(uri).await {
+            Ok(resp) => resp,
+            Err(e) => {
+                error!("Error requested uri: {endpoint} {e}.");
+                bail!("Error requested uri: {endpoint}.")
+            }
+        };
+
+        let bytes = match body::to_bytes(resp.into_body()).await {
+            Ok(bytes) => bytes,
+            Err(e) => {
+                error!("Error read data from upstream: {endpoint} {e}.");
+                bail!("Error read data from upstream: {endpoint}.")
+            }
+        };
+
+        let indexer_response: IndexerResponse = match from_slice(&bytes) {
+            Ok(indexer_response) => indexer_response,
+            Err(e) => {
+                error!("Error parsed indexer response from upstream: {endpoint} {e}.");
+                bail!("Error parsed indexer response from upstream: {endpoint}.")
+            }
+        };
+
+        debug!("Received indexer response for {cid}: {indexer_response:?}.");
+
+        choose_provider(indexer_response)
+
+        // TODO:
+        // 1. filter FleekNetwork metadata
+        // 2. pick node (round-robin)
+        // 3. call get_block to node
+        // 4.
+        //   4.1 return block?
+        //   4.2 resolve?
+        //
+        // IMPROVEMENTS:
+        // 1. maintain N workers keep track of indexing data
+        // 2. cherry-pick closest node
+        // 3. cache TTL
+    }
+
+    pub async fn resolve_content(&self, addrs: Vec<SocketAddrV4>, cid: &str) -> Result<Vec<u8>> {
```

Comment:
I am guessing this is just temp, as this function will be replaced and we should support streaming

Reply:
Yup 💯